src/v/cluster/producer_state.h

```diff
@@ -0,0 +1,235 @@
+struct test_fixture;
+
+namespace cluster {
+
+template<class Func>
+concept AcceptsUnits = requires(Func f, ssx::semaphore_units units) {
+    f(std::move(units));
+};
+
+class producer_state_manager;
+class producer_state;
+struct producer_state_snapshot;
+class request;
+
+using producer_ptr = ss::lw_shared_ptr<producer_state>;
+// Note: A shared_promise doesn't guarantee available() to be true
+// right after set_value(), this is an implementation quirk, be
+// mindful of that behavior when using it. We have a test for
+// it in expiring_promise_test
+using result_promise_t = ss::shared_promise<result<kafka_result>>;
+using request_ptr = ss::lw_shared_ptr<request>;
+using seq_t = int32_t;
+
+/// A request for a given sequence range, both inclusive.
+/// The sequence numbers are stamped by the client and are a part
+/// of batch header. A request can either be in progress or completed
+/// depending on the whether the holding promise is set.
+class request {
+public:
+    explicit request(
+      seq_t first, seq_t last, model::term_id term, result_promise_t res)
+      : _first_sequence(first)
+      , _last_sequence(last)
+      , _term(term)
+      , result(std::move(res)) {}
+
+    template<class ValueType>
+    void set_value(ValueType&& value) {
+        result.set_value(std::forward<ValueType>(value));
+    }
+
+    bool operator==(const request&) const;
+
+private:
+    seq_t _first_sequence;
+    seq_t _last_sequence;
+    // term in which the request was submitted.
+    model::term_id _term;
+    // Set when the result for this request is finished. This is a shared
+    // promise because a client can retry an already in progress request
+    // (eg: timeouts) and we just chain the retried request request with
+    // with the future from an already in-flight promise with the sequence
+    // number match.
+    result_promise_t result;
+
+    bool in_progress() const { return !result.available(); }
+    friend class requests;
+    friend class producer_state;
+};
+
+// A cached buffer of requests, the requests can be in progress / finished.
+// A request is promoted from inflight to finished once it is applied in the
+// log.
+//
+// We retain a maximum of `requests_cached_max` finished requests.
+// Kafka clients only issue requests in batches of 5, the queue is fairly small
+// at all times.
+class requests {
+public:
+    using request_result_t
+      = std::variant<errc, ss::future<result<kafka_result>>, request_ptr>;
+    request_result_t
+    new_request(seq_t first, seq_t last, model::term_id current);
+
+    bool stm_apply(const model::batch_identity& bid, kafka::offset offset);
+
+    void shutdown();
+
+    bool operator==(const requests&) const;
+    friend std::ostream& operator<<(std::ostream&, const requests&);
+
+private:
+    static constexpr int32_t requests_cached_max = 5;
+    std::optional<request_ptr> last_request() const;
+    ss::chunked_fifo<request_ptr> _inflight_requests;
```

Comment:
as we chated yesterday since the collections in here are bounded to 5 elements it seems wasteful to use `chunked_fifo`, i think vector would be a better fit

Reply:
sounds great 👍🏼